### ISSUE TITLE
Proper way of fitting classifiers before creating an heterogeneous pool

### ISSUE THREAD

**Issue by francescopisu** [scikit-learn-contrib/DESlib#264]
Hey, I'm working on a research paper focused on building a binary classification model in the biomedical domain. The dataset comprises approximately 800 data points. By following the instructions on the examples, I've found two different ways of splitting the dataset and fitting the base classifiers of the pool.

1) Split in train/test (e.g., 75/25) and then split the training in train/dsel (e.g., 50/50). 
In [this random forest example](https://deslib.readthedocs.io/en/latest/auto_examples/plot_random_forest.html#sphx-glr-auto-examples-plot-random-forest-py), the RF is fitted on the 75% training portion and the DS methods on the 50% DSEL portion. 
2) In all the other examples, the 50% training portion is used to fit the classifier and the 50% DSEL portion is used to fit DS methods.

Furthermore, I wanted to point out this tip taken from the [tutorial](https://deslib.readthedocs.io/en/latest/user_guide/tutorial.html#running-dynamic-selection-with-bagging) : 

> An important point here is that in case of small datasets or when the base classifier models in the pool are weak estimators such as Decision Stumps or Perceptrons, an overlap between the training data and DSEL may be beneficial for achieving better performance.

That seems my case, as my dataset is rather small compared to the standard datasets. Hence, I was thinking of fitting my base classifiers on the 75% part and then leveraging some overlap to get better performance (and this is really the case! In fact, overlapping leads to a median auc of 0.76 whereas non-overlapping gives 0.71).

What would be the best way of dealing with the problem ?

**Comment by francescopisu** [comment]
Any update on this ?

**Comment by Menelau** [comment]
Hello,

In you case (small dataset with about 800 samples), I do believe that using the whole 75% for training the base models and DSEL I'm pretty sure the best way of handling it is having some overlap between the datasets.

Does your dataset also suffers from class imbalance? If yes, this is another indication that you should consider an overlapping approach as the minority class may be quite under represented if we divide the dataset too much. If have done that in quite a few papers such as the FIRE-DES++ one:

 FIRE-DES++: Enhanced online pruning of base classifiers for dynamic ensemble selection: https://www.etsmtl.ca/Unites-de-recherche/LIVIA/Recherche-et-innovation/Publications/Publications-2019/Cruz_PR_2019.pdf

 Also, if your dataset suffer from class imbalance, maybe you could try applying data augmentation to increase the DSEL size later as we did in a previous publication:

A study on combining dynamic selection and data preprocessing for imbalance learning: https://en.etsmtl.ca/Unites-de-recherche/LIVIA/Recherche-et-innovation/Publications/Publications-2018/Roy_Neurocomputing_2018_InPress.pdf

**Comment by francescopisu** [comment]
Hello Menelau, thank you very much for your time and valuable inputs. My dataset suffers from mild imbalance, but this time is the negative class (let's say absence of condition) to be underrepresented, which is the class I'm less interested in. I'm going to read these two articles and try to apply what you suggested.

**Comment by francescopisu** [comment]
I have some feedback regarding the overlap. If I understood correctly, Xtrain and Xdsel should be taken such that set.intersection(Xnew_train, Xdsel) is not empty set.

Let Xtrain, Xtest = train_test_split(features, target, test_size=0.2) 
be the train and test set after the first split (test set is 20% of the original dataset).

Then, I'm sampling 80% of entries from Xtrain to get the new training set Xnew_train. The remaining 20% is the Xdsel set. As of now there's no overlap.

To introduce overlap, I'm randomly sampling 10% of the entries in Xnew_train and adding them to Xdsel to get Xnew_dsel. 

Example with easy numbers:
dataset = (1000, n_features)
Xtrain, Xtest = (800, n_features), (200, n_features)
Xnew_train = 80% of 800 -> (640, n_features)
Xdsel = 20% of 800 -> (160, n_features)
Xnew_dsel = take 10% from Xnew_train: 64 -> (64+160=224, n_features)

I'm also doing it in the cross-validation procedure: I take 10% from the training fold to add it to the validation portion. 

I think I'm getting very biased results:
| Metric | Training | Validation |
| ------ | -------- | ---------- | 
| sensitivity|0.893 - [0.85-0.92]|0.987 - [0.95-1.0]
specificity|0.512 - [0.43-0.62]|0.741 - [0.64-0.87]
fpr|0.488 - [0.38-0.57]|0.259 - [0.13-0.36]
fnr|0.107 - [0.08-0.15]|0.013 - [0.0-0.05]
ppv|0.766 - [0.74-0.8]|0.87 - [0.82-0.93]
npv|0.728 - [0.68-0.79]|0.972 - [0.89-1.0]
f1|0.826 - [0.81-0.84]|0.924 - [0.89-0.96]
f2|0.866 - [0.84-0.89]|0.962 - [0.93-0.98]
auc|0.843 - [0.82-0.87]|0.884 - [0.81-0.94]
brier|0.242 - [0.22-0.27]|0.101 - [0.06-0.15]

Am I implementing the overlapping correctly ?

Thank you for your time
